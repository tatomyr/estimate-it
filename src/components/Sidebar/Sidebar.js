import React from 'react'
import PropTypes from 'prop-types'
import SideButton from '../SideButton'
import { matchType, estimateType } from '../../helpers/propTypes'

const Sidebar = ({
  match: { params: { estimateId = 'new' } },
  recalc,
  saveEstimate,
  estimates,
  username,
  graphView,
  minifyGraph,
  enlargeGraph,
}) => {
  const { saved, calculated } = estimates[estimateId]
  return (
    <aside>
      <SideButton
        title="New"
        name="file"
        disabled={estimateId === 'new'}
        link="/estimate/new"
      />
      <SideButton
        title="Calculate"
        name="rocket"
        disabled={calculated}
        onClick={() => !calculated && recalc(estimateId)}
      />
      <SideButton
        title="Save"
        name="save"
        disabled={!username || saved}
        color={(saved && 'secondary') || (username && calculated && 'primary') || 'warning'}
        onClick={() => !saved && saveEstimate({ estimateId })}
      />
      {graphView === 'minified' ? (
        <SideButton
          title="Graph"
          name="bar-chart"
          onClick={enlargeGraph}
        />
      ) : (
        <SideButton
          title="Editor"
          name="edit"
          onClick={minifyGraph}
        />
      )}
      <div className="sidebar-padding" />
      <SideButton
        title="Dashboard"
        name="list-ul"
        link="/dashboard"
      />
      <SideButton
        title="Auth"
        name={username ? 'unlock' : 'lock'}
        color={username ? 'success' : 'warning'}
        link="/auth"
      />
      <SideButton
        title="Home"
        name="home"
        link="/"
      />
    </aside>
  )
}

Sidebar.propTypes = {
  match: matchType.isRequired,
  recalc: PropTypes.func.isRequired,
  saveEstimate: PropTypes.func.isRequired,
  estimates: PropTypes.objectOf(estimateType).isRequired,
  username: PropTypes.string.isRequired,
  graphView: PropTypes.string.isRequired,
  minifyGraph: PropTypes.func.isRequired,
  enlargeGraph: PropTypes.func.isRequired,
}

export default Sidebar
