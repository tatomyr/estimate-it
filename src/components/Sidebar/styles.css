aside {
  width: 48px;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 4px 0;
  background-color: #222;
  color: #fff;
}
