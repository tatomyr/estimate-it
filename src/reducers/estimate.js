import { handleFlat } from '../helpers/task-parsing'
import defaultText from '../helpers/default-text'
import { CHANGE_TEXT, RECALCULATE } from '../actions/types'

const defaultState = {
  text: defaultText,
  graphData: [],
}

export default (state = defaultState, action) => {
  switch (action.type) {
    case CHANGE_TEXT:
      return ({
        ...state,
        text: action.text,
      })
    case RECALCULATE:
      return ({
        ...state,
        ...handleFlat(state.text, 15),
      })
    default:
      return state
  }
}
