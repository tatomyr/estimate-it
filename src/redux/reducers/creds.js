import { SET_CREDS, RESET_CREDS } from '../actions/types'

export const emptyCreds = ({
  haveBeenChecked: false,
  username: '',
  dbName: '',
  apiKey: '',
})

export default (state = emptyCreds, { type, payload }) => {
  switch (type) {
    case SET_CREDS:
      return ({
        ...payload,
        haveBeenChecked: true,
      })

    case RESET_CREDS:
      return ({
        ...emptyCreds,
        haveBeenChecked: true,
      })

    default:
      return state
  }
}
