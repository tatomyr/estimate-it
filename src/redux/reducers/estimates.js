import { handleText, parseParam } from '../../helpers/tasks-parsing'
import {
  UPDATE_ESTIMATE,
  RECALCULATE,
  CLEAN_ESTIMATE,
  CLEAN_ALL_ESTIMATES,
  SET_TITLES,
  MARK_ESTIMATE_SAVED,
} from '../actions/types'

export const emptyEstimate = ({
  _id: 'new',
  text: '',
  graphData: [],
  project: '',
  participants: [],
  calculated: true,
  saved: true,
})

const getProjectName = text => parseParam(text)('@project')
const getParticipants = text => parseParam(text)('@participants')
  .split(',')
  .map(participant => participant.trim())
  .filter(participant => participant)

export default (state = ({ new: emptyEstimate }), { type, payload }) => {
  switch (type) {
    case UPDATE_ESTIMATE:
    {
      const { _id, text } = payload
      return ({
        ...state,
        [_id]: {
          ...state[_id],
          graphData: [],
          calculated: false,
          saved: false,
          project: getProjectName(text),
          participants: getParticipants(text),
          ...payload,
        },
      })
    }

    case CLEAN_ESTIMATE:
    {
      // Putting 'new' here to avoid creating an estimate with an undefined `_id`
      // ...while logging out on different routes
      // ...(that may not contain an `:estimateId` param)
      const { _id = 'new' } = payload
      return ({
        ...state,
        [_id]: {
          ...emptyEstimate,
          _id,
        },
      })
    }

    case CLEAN_ALL_ESTIMATES:
      return ({
        new: emptyEstimate,
      })

    case RECALCULATE:
    {
      const { _id } = payload
      return ({
        ...state,
        [_id]: {
          ...state[_id],
          _id,
          ...handleText(state[_id].text),
          calculated: true,
          saved: false,
        },
      })
    }

    case SET_TITLES:
    {
      const { projects } = payload
      return ({
        ...state,
        ...projects.reduce(($, project) => ({
          ...$,
          [project._id]: {
            ...project,
            saved: true,
          },
        }), {}),
      })
    }

    case MARK_ESTIMATE_SAVED:
    {
      const { _id } = payload
      return ({
        ...state,
        [_id]: {
          ...state[_id],
          saved: true,
        },
      })
    }

    default:
      return state
  }
}
