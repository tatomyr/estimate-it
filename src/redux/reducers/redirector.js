import { SET_LOCATION, RESET_LOCATION } from '../actions/types'

export default (state = ({
  location: '',
}), { type, payload }) => {
  switch (type) {
    case SET_LOCATION:
      return ({
        location: payload,
      })

    case RESET_LOCATION:
      return ({
        location: '',
      })

    default:
      return state
  }
}
