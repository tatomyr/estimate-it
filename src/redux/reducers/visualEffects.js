import {
  ADD_SPINNER,
  DEL_SPINNER,
  ENLARGE_GRAPH,
  MINIFY_GRAPH,
} from '../actions/types'

export default (state = {
  spinnersCount: false,
  graphView: 'minified',
}, { type }) => {
  switch (type) {
    case ADD_SPINNER:
      return ({
        ...state,
        spinnersCount: state.spinnersCount + 1,
      })

    case DEL_SPINNER:
      return ({
        ...state,
        spinnersCount: state.spinnersCount - 1,
      })

    case ENLARGE_GRAPH:
      return ({
        ...state,
        graphView: 'full',
      })

    case MINIFY_GRAPH:
      return ({
        ...state,
        graphView: 'minified',
      })

    default:
      return state
  }
}
