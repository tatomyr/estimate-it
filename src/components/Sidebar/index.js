import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import Sidebar from './Sidebar'
import './styles.css'
import {
  recalc,
  minifyGraph,
  enlargeGraph,
} from '../../redux/actions'
import {
  saveEstimate,
} from '../../redux/actions/async'

const mapStateToProps = ({
  estimates,
  creds: { username },
  visualEffects: { graphView },
}) => ({
  estimates,
  username,
  graphView,
})

const mapDispatchToProps = ({
  recalc,
  saveEstimate,
  minifyGraph,
  enlargeGraph,
})

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Sidebar))
