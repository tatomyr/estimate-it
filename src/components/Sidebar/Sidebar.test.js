/* globals describe, it, expect */

import React from 'react'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import Sidebar from './Sidebar'
import { emptyEstimate } from '../../redux/reducers/estimates'

describe('Sidebar', () => {
  it('renders correctly', () => {
    configure({ adapter: new Adapter() })
    const tree = shallow(<Sidebar
      match={{ params: { estimateId: 'new' } }}
      location={{ pathname: '/' }}
      recalc={() => null}
      saveEstimate={() => null}
      estimates={{ new: emptyEstimate }}
      username="Test User"
      graphView="minified"
      minifyGraph={() => null}
      enlargeGraph={() => null}
    />, {})
    expect(tree).toMatchSnapshot()
  })
})
