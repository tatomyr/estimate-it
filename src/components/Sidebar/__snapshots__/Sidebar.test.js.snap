// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Sidebar
    enlargeGraph={[Function]}
    estimates={
        Object {
            "new": Object {
              "_id": "new",
              "calculated": true,
              "graphData": Array [],
              "participants": Array [],
              "project": "",
              "saved": true,
              "text": "",
            },
          }
    }
    graphView="minified"
    location={
        Object {
            "pathname": "/",
          }
    }
    match={
        Object {
            "params": Object {
              "estimateId": "new",
            },
          }
    }
    minifyGraph={[Function]}
    recalc={[Function]}
    saveEstimate={[Function]}
    username="Test User"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Connect(SideButton)
          disabled={true}
          link="/estimate/new"
          name="file"
          title="New"
/>,
        <Connect(SideButton)
          disabled={true}
          name="rocket"
          onClick={[Function]}
          title="Calculate"
/>,
        <Connect(SideButton)
          color="secondary"
          disabled={true}
          name="save"
          onClick={[Function]}
          title="Save"
/>,
        <Connect(SideButton)
          name="bar-chart"
          onClick={[Function]}
          title="Graph"
/>,
        <div
          className="sidebar-padding"
/>,
        <Connect(SideButton)
          link="/dashboard"
          name="list-ul"
          title="Dashboard"
/>,
        <Connect(SideButton)
          color="success"
          link="/auth"
          name="unlock"
          title="Auth"
/>,
        <Connect(SideButton)
          link="/"
          name="home"
          title="Home"
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "disabled": true,
          "link": "/estimate/new",
          "name": "file",
          "title": "New",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "disabled": true,
          "name": "rocket",
          "onClick": [Function],
          "title": "Calculate",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "color": "secondary",
          "disabled": true,
          "name": "save",
          "onClick": [Function],
          "title": "Save",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "name": "bar-chart",
          "onClick": [Function],
          "title": "Graph",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "sidebar-padding",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "link": "/dashboard",
          "name": "list-ul",
          "title": "Dashboard",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "color": "success",
          "link": "/auth",
          "name": "unlock",
          "title": "Auth",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "link": "/",
          "name": "home",
          "title": "Home",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "aside",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Connect(SideButton)
            disabled={true}
            link="/estimate/new"
            name="file"
            title="New"
/>,
          <Connect(SideButton)
            disabled={true}
            name="rocket"
            onClick={[Function]}
            title="Calculate"
/>,
          <Connect(SideButton)
            color="secondary"
            disabled={true}
            name="save"
            onClick={[Function]}
            title="Save"
/>,
          <Connect(SideButton)
            name="bar-chart"
            onClick={[Function]}
            title="Graph"
/>,
          <div
            className="sidebar-padding"
/>,
          <Connect(SideButton)
            link="/dashboard"
            name="list-ul"
            title="Dashboard"
/>,
          <Connect(SideButton)
            color="success"
            link="/auth"
            name="unlock"
            title="Auth"
/>,
          <Connect(SideButton)
            link="/"
            name="home"
            title="Home"
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "disabled": true,
            "link": "/estimate/new",
            "name": "file",
            "title": "New",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "disabled": true,
            "name": "rocket",
            "onClick": [Function],
            "title": "Calculate",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "color": "secondary",
            "disabled": true,
            "name": "save",
            "onClick": [Function],
            "title": "Save",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "name": "bar-chart",
            "onClick": [Function],
            "title": "Graph",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "sidebar-padding",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "link": "/dashboard",
            "name": "list-ul",
            "title": "Dashboard",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "color": "success",
            "link": "/auth",
            "name": "unlock",
            "title": "Auth",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "link": "/",
            "name": "home",
            "title": "Home",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "aside",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
